### DISCUSSION TITLE
Integrating `wry` into pre-existing GTK project

### DISCUSSION THREAD

**Discussion by emmyoh** [tauri-apps/wry#1178] [General]
Hello,

I currently have a pre-existing GTK 4 project (a Web browser that I'm building) using [WebKit 6.0](https://gitlab.gnome.org/World/Rust/webkit6-rs). I'm interested in switching over to `wry` for the following reasons:
1. Better multi-platform support; at the moment, my project only builds on Linux
2. Usage of [`with_asynchronous_custom_protocol`](https://docs.rs/wry/latest/wry/struct.WebViewBuilder.html#method.with_asynchronous_custom_protocol); custom protocols using the WebKit Rust bindings cannot have asynchronous responders AFAIK. **This is my biggest reason for wanting to switch**. By the way, how did you implement this yourselves with the WebKit backend?
3. Interest in eventually moving to Servo with ease once it's supported with `wry`.

What would I need to rip out and replace in my GTK app? Does switching over to `wry` require fundamentally redoing how I'm using my app's GUI framework, or is it as simple as swapping out the WebKitGTK portion of my code with some sort of `wry` GTK component?

**Comment by amrbashir** [comment]
Just a disclaimer, we are currently using GTK 3 not GTK 4 and don't have an plans atm to switch to GTK 4



> * Usage of [`with_asynchronous_custom_protocol`](https://docs.rs/wry/latest/wry/struct.WebViewBuilder.html#method.with_asynchronous_custom_protocol); custom protocols using the WebKit Rust bindings cannot have asynchronous responders AFAIK. **This is my biggest reason for wanting to switch**. By the way, how did you implement this yourselves with the WebKit backend?
> * Interest in eventually moving to Servo with ease once it's supported with `wry`.

We are using [`finish_with_response`](https://docs.rs/webkit2gtk/latest/webkit2gtk/trait.URISchemeRequestExt.html#method.finish_with_response) but IIRC, even if is asynchronous, it still need to be called on the main thread.

> What would I need to rip out and replace in my GTK app? Does switching over to wry require fundamentally redoing how I'm using my app's GUI framework, or is it as simple as swapping out the WebKitGTK portion of my code with some sort of wry GTK component?

Depends on your current GUI requirements, WRY doesn't provide any components other than a webview, so you will need another GUI crate.

If your app is mainly a webview, you can just use `tao` crate as a windowing crate and initialize the WRY webview inside its windows.

if you need more than a webview then it's up to you to pick and choose from either using `winit` and one of these rendering crates:
- glutin
- wgpu

Or a full-feature UI crate:
- egui
- freya
- slint
- iced
- druid

Also take a look at https://areweguiyet.com/

And here is an example of integrating with `winit + wgpu` https://github.com/tauri-apps/wry/blob/dev/examples/wgpu.rs
